ipc/provider/src/manager/evm/manager.rs

```diff
@@ -1076,6 +1076,37 @@ impl BottomUpCheckpointRelayer for EthSubnetManager {
 
         Ok(events)
     }
+
+    async fn max_quorum_reached_height(
+        &self,
+        subnet_id: &SubnetID,
+        from: ChainEpoch,
+        to: ChainEpoch,
+    ) -> Result<Option<ChainEpoch>> {
+        let mut maybe_height = None;
+
+        let period = self.checkpoint_period(subnet_id).await?;
+
+        let mut l = from;
+        let mut r = to;
+
+        while l < r {
+            let mid = (l + r) / 2;
+
+            // locate the epoch where checkpoints should have been created
+            let mid = mid / period * period;
+
+            if self.quorum_reached_events(mid).await?.is_empty() {
+                r = mid - 1;
+            } else {
+                maybe_height = Some(mid);
+                l = mid + 1;
+            }
+        }
+
+        Ok(maybe_height)
```

Comment:
It may be worth factoring out the bisection algorithm and make some unit tests, because it's not a standard one and looks tricky. Not sure what are the invariants it's trying to adhere to. 

I may be wrong but it looks to me like it could get into an infinite loop: 
1. Say our check period is 10, and we call this method with `from = 53` and `to = 62`. 
2. `mid = (53+62)/2 = 57`
3. `mid = mid / 10 * 10` -> I'm not sure what this should do; mathematically it should be 57, but it might give 50 due to truncation. Let's say after this `mid = 50`; that's outside `from` and `to`. Is that okay for the method to return?

4. Say `quorum_reached_events(50)` is _not_ empty. `maybe_height = 50`, and `from = 51`. 
5. Goto 2. `mid = (51+62)/2 = 56`, and we'll enter another loop with `mid = 50`. 

4. Say `quorum_reached_events(50)` is empty. in this case `to = 49` and we exit with an empty result. Maybe there would have been a result at 60?

I'm also not sure what why we're looking for quorum at fixed multiples of the period? Signatures are submitted asynchronously, they can gather at any height. I'm not even sure why we'd be looking for checkpoints added to the ledger at fixed multiples of the period, since batching could trigger a checkpoint at any height.

Reply:
yeah, there is a `max-pending` flag now, I dont think this would be needed anymore. Might be difficult to add tests for this one too.